controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java

```diff
@@ -189,52 +203,67 @@ public void cancelChange(ApplicationId applicationId, boolean keepApplicationCha
                                      .filter(JobError.outOfCapacity::equals).isPresent();
         if (isRetry) reason += "; retrying on out of capacity";
 
-        Change change = application.change();
-        // For both versions, use the newer of the change's and the currently deployed versions, or a fallback if none of these exist.
-        Version platform = jobType == JobType.component
-                ? Version.emptyVersion
-                : deploymentFor(application, jobType).map(Deployment::version)
-                                                     .filter(version -> ! change.upgrades(version))
-                                                     .orElse(change.platform()
-                                                                   .orElse(application.oldestDeployedPlatform()
-                                                                                      .orElse(controller.systemVersion())));
-        ApplicationVersion applicationVersion = jobType == JobType.component
-                ? ApplicationVersion.unknown
-                : deploymentFor(application, jobType).map(Deployment::applicationVersion)
-                                                     .filter(version -> ! change.upgrades(version))
-                                                     .orElse(change.application()
-                                                                   .orElseGet(() -> application.oldestDeployedApplication()
-                                                                                               .orElseThrow(() -> new IllegalArgumentException("Cannot determine application version to use for " + jobType))));
-
-        return new Job(application, jobType, reason, availableSince, concurrentlyWith, isRetry, change, platform, applicationVersion);
+        return new Job(application, target, jobType, reason, availableSince, concurrentlyWith, isRetry, change.application().isPresent());
+    }
+
+    private Version targetPlatform(Application application, Change change, Optional<Deployment> deployment) {
+        return max(deployment.map(Deployment::version), change.platform())
+                .orElse(application.oldestDeployedPlatform()
+                                   .orElse(controller.systemVersion()));
+    }
+
+    private ApplicationVersion targetApplication(Application application, Change change, Optional<Deployment> deployment) {
+        return max(deployment.map(Deployment::applicationVersion), change.application())
+                .orElse(application.oldestDeployedApplication()
+                                   .orElse(application.deploymentJobs().jobStatus().get(component).lastSuccess().get().applicationVersion()));
+    }
+
+    private static <T extends Comparable<T>> Optional<T> max(Optional<T> o1, Optional<T> o2) {
+        return ! o1.isPresent() ? o2 : ! o2.isPresent() ? o1 : o1.get().compareTo(o2.get()) >= 0 ? o1 : o2;
```

Comment:
Hmm... I read this very comfortably as 
_if o2 is empty, return o1; if o1 is empty, return o2; otherwise, return the maximum among them._
I won't argue the Stream approach is hard to read, but I generally prefer code with less overhead. I can use Stream if you still want me to.